### PR TITLE
Fix Python SDK event deserialization key mismatches (snake_case → camelCase)

### LINKED ISSUE #21713: Python SDK: StepEventMetadata.from_json uses wrong key for detailed_diff (snake_case vs camelCase)
### What happened?

`StepEventMetadata.from_json()` in the Python Automation API reads `data.get("detailed_diff")` (snake_case), but the Go engine serializes this field as `"detailedDiff"` (camelCase). This means `meta.detailed_diff` is **always `None`** when consuming engine events, regardless of whether the provider returns detailed diffs.

A secondary issue: even if the key matched, the raw dict values are passed through without being deserialized into `PropertyDiff` objects.

### Root cause

In [`sdk/python/lib/pulumi/automation/events.py` (line 461)](https://github.com/pulumi/pulumi/blob/ab9ad03ebb2d/sdk/python/lib/pulumi/automation/events.py#L461):

```python
detailed_diff=data.get("detailed_diff"),  # snake_case — never matches
```

The Go engine struct in [`sdk/go/common/apitype/events.go` (line 210)](https://github.com/pulumi/pulumi/blob/ab9ad03ebb2d/sdk/go/common/apitype/events.go#L210) has:

```go
DetailedDiff map[string]PropertyDiff `json:"detailedDiff"`  // camelCase
```

### This is clearly an oversight

Every other field in the same `from_json` method correctly uses camelCase matching the Go JSON tags:

| Python `data.get(...)` | Go `json:` tag | Correct? |
|---|---|---|
| `"streamId"` | `"streamID"` | ✅ |
| `"maybeCorrupt"` | `"maybeCorrupt"` | ✅ |
| `"durationSeconds"` | `"durationSeconds"` | ✅ |
| `"resourceChanges"` | `"resourceChanges"` | ✅ |
| `"retainOnDelete"` | `"retainOnDelete"` | ✅ |
| `"initErrors"` | `"initErrors"` | ✅ |
| `"enforcementLevel"` | `"enforcementLevel"` | ✅ |
| **`"detailed_diff"`** | **`"detailedDiff"`** | ❌ |

### Same bug in PolicyEvent

[`PolicyEvent.from_json()` (line 197)](https://github.com/pulumi/pulumi/blob/ab9ad03ebb2d/sdk/python/lib/pulumi/automation/events.py#L197) has the same issue:

```python
resource_urn=data.get("resource_urn"),  # snake_case — never matches
```

Go tag ([line 57](https://github.com/pulumi/pulumi/blob/ab9ad03ebb2d/sdk/go/common/apitype/events.go#L57)): `json:"resourceUrn,omitempty"`

### Suggested fix

```python
# StepEventMetadata.from_json (line 461)
- detailed_diff=data.get("detailed_diff"),
+ detailed_diff={
+     k: PropertyDiff.from_json(v) if isinstance(v, dict) else v
+     for k, v in raw_dd.items()
+ } if (raw_dd := data.get("detailedDiff")) else None,

# PolicyEvent.from_json (line 197)
- resource_urn=data.get("resource_urn"),
+ resource_urn=data.get("resourceUrn"),
```

### Impact

Anyone using the Python Automation API to consume engine events (e.g. `on_event` callbacks with `preview()` or `up()`) and trying to read `meta.detailed_diff` gets `None` and must fall back to manually diffing `old.inputs` vs `new.inputs`. This affects property-level diff display in tools built on the Automation API.

### Output of `pulumi about`

```
SDK: Python 3.219.0
```

This bug appears to have been present since `detailed_diff` was first added to the Python SDK.

### PR DESCRIPTION
**Pull request by JayDoubleu**
## Summary

- Fix `StepEventMetadata.from_json()` reading `"detailed_diff"` (snake_case) instead of `"detailedDiff"` (camelCase) — the key emitted by the Go engine. This caused `meta.detailed_diff` to always be `None`.
- Additionally deserialize raw dicts into `PropertyDiff` objects via `PropertyDiff.from_json()`.
- Fix `PolicyEvent.from_json()` reading `"resource_urn"` instead of `"resourceUrn"` — same snake_case vs camelCase mismatch.
- Add unit tests covering both fixes and verifying that old snake_case keys are no longer read.

Every other field in these `from_json` methods already uses camelCase matching the Go JSON tags (`"streamId"`, `"maybeCorrupt"`, `"retainOnDelete"`, etc.) — these two were the only outliers.

Fixes #21713

## Test plan

- [x] New unit tests in `sdk/python/lib/test/automation/test_events.py` (6 tests, all passing)
- [ ] Existing automation test suite (`test_local_workspace.py::test_structured_events`) should continue to pass

### CODE REVIEW

File: sdk/python/lib/test/automation/test_events.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            "color": "",
+            "policyName": "policy",
+            "policyPackName": "pack",
+            "policyPackVersion": "1",
+            "policyPackVersionTag": "v1",
+            "enforcementLevel": "mandatory",
+            "resource_urn": "urn:pulumi:stack::project::type::name",
+        }
+        event = PolicyEvent.from_json(data)
+        self.assertIsNone(event.resource_urn)
```
Comment (Frassle) [inline sdk/python/lib/test/automation/test_events.py]: Same as above

File: sdk/python/lib/test/automation/test_events.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            "op": "update",
+            "urn": "urn:pulumi:stack::project::type::name",
+            "type": "type",
+            "provider": "provider",
+            "detailed_diff": {
+                "tags": {"diffKind": "update", "inputDiff": False},
+            },
+        }
+        meta = StepEventMetadata.from_json(data)
+        self.assertIsNone(meta.detailed_diff)
```
Comment (Frassle) [inline sdk/python/lib/test/automation/test_events.py]: Probably don't need to test the negative here

File: sdk/python/lib/pulumi/automation/events.py [hunk 3]
```diff
@@ -458,7 +458,12 @@ def from_json(cls, data: dict) -> "StepEventMetadata":
             new=StepEventStateMetadata.from_json(new) if new else None,
             keys=data.get("keys"),
             diffs=data.get("diffs"),
-            detailed_diff=data.get("detailed_diff"),
+            detailed_diff={
+                k: PropertyDiff.from_json(v) if isinstance(v, dict) else v
```
Comment (Frassle) [inline sdk/python/lib/pulumi/automation/events.py]: Isn't 'v' always a dict?

**Comment by JayDoubleu** [comment]
Thanks for the review @Frassle! Addressed all three comments:

1. **`isinstance(v, dict)` guard**: Removed — `PropertyDiff` is always a JSON object from Go, so `v` is always a dict.
2. **Negative tests**: Removed both `test_snake_case_key_not_read` tests.

Force-pushed with the fixes.

**Comment by Frassle** [comment]
/run-acceptance-tests
Please view the results of the acceptance tests [Here](https://github.com/pulumi/pulumi/actions/runs/21787058151)

File: sdk/python/lib/pulumi/automation/events.py [hunk 4]
```diff
@@ -458,7 +458,9 @@ def from_json(cls, data: dict) -> "StepEventMetadata":
             new=StepEventStateMetadata.from_json(new) if new else None,
             keys=data.get("keys"),
             diffs=data.get("diffs"),
-            detailed_diff=data.get("detailed_diff"),
+            detailed_diff={k: PropertyDiff.from_json(v) for k, v in raw_dd.items()}
+            if (raw_dd := data.get("detailedDiff"))
+            else None,
```
Comment (tgummerer) [inline sdk/python/lib/pulumi/automation/events.py]: Why reassign raw_dd here? Wouldn't this need to be in the line above?  Also I don't think you need the if here.

File: sdk/python/lib/test/automation/test_events.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+class TestStepEventMetadataFromJson(unittest.TestCase):
+    """Test StepEventMetadata.from_json reads camelCase keys matching Go JSON tags."""
+
+    def test_detailed_diff_camel_case_key(self):
+        """detailedDiff (camelCase from Go engine) should be deserialized."""
+        data = {
+            "op": "update",
+            "urn": "urn:pulumi:stack::project::type::name",
+            "type": "type",
+            "provider": "provider",
```
Comment (tgummerer) [inline sdk/python/lib/test/automation/test_events.py]: You're not testing this part, so you can omit it. Same below.

File: sdk/python/lib/test/automation/test_events.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    def test_detailed_diff_none_when_absent(self):
+        """detailed_diff should be None when detailedDiff is not in the data."""
+        data = {
+            "op": "create",
+            "urn": "urn:pulumi:stack::project::type::name",
+            "type": "type",
+            "provider": "provider",
+        }
+        meta = StepEventMetadata.from_json(data)
+        self.assertIsNone(meta.detailed_diff)
```
Comment (tgummerer) [inline sdk/python/lib/test/automation/test_events.py]: This test doesn't seem very useful.

File: sdk/python/lib/test/automation/test_events.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            "op": "update",
+            "urn": "urn:pulumi:stack::project::type::name",
+            "type": "type",
+            "provider": "provider",
+            "detailedDiff": {
+                "tags": {"diffKind": "update", "inputDiff": False},
+            },
+        }
+        meta = StepEventMetadata.from_json(data)
+        self.assertIsInstance(meta.detailed_diff["tags"], PropertyDiff)
```
Comment (tgummerer) [inline sdk/python/lib/test/automation/test_events.py]: JUst test this in the test above instead of addin ganother test?  Also you can probably just assert on the whole shape of the detailed diff.

File: sdk/python/lib/test/automation/test_events.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    def test_resource_urn_camel_case_key(self):
+        """resourceUrn (camelCase from Go engine) should be deserialized."""
+        data = {
+            "message": "test",
+            "color": "",
+            "policyName": "policy",
+            "policyPackName": "pack",
+            "policyPackVersion": "1",
+            "policyPackVersionTag": "v1",
+            "enforcementLevel": "mandatory",
```
Comment (tgummerer) [inline sdk/python/lib/test/automation/test_events.py]: This shouldn't be necessary.

Comment (JayDoubleu) [reply sdk/python/lib/test/automation/test_events.py]: Addressed all the feedback, thanks! Just to double check on this one.. did you mean drop the PolicyEvent test entirely? I removed it but since this PR also fixes the `resource_urn` -> `resourceUrn` key in `PolicyEvent.from_json`, wanted to make sure you're happy with no test coverage for that part.

File: sdk/python/lib/pulumi/automation/events.py [hunk 9]
```diff
@@ -458,7 +459,9 @@ def from_json(cls, data: dict) -> "StepEventMetadata":
             new=StepEventStateMetadata.from_json(new) if new else None,
             keys=data.get("keys"),
             diffs=data.get("diffs"),
-            detailed_diff=data.get("detailed_diff"),
+            detailed_diff={k: PropertyDiff.from_json(v) for k, v in raw_dd.items()}
+            if raw_dd
+            else None,
```
Comment (tgummerer) [inline sdk/python/lib/pulumi/automation/events.py]: Why do we need this if?  I'm still confused by this.

File: sdk/python/lib/test/automation/test_events.py (hunk 8 above)
Comment (tgummerer) [reply sdk/python/lib/test/automation/test_events.py]: I meant just the data I commented on here, we just need the `resourceUrn` bit we should still have the test.

File: sdk/python/lib/pulumi/automation/events.py (hunk 9 above)
Comment (JayDoubleu) [reply sdk/python/lib/pulumi/automation/events.py]: Fixed, switched to `data.get("detailedDiff", {})`

File: sdk/python/lib/test/automation/test_events.py (hunk 8 above)
Comment (JayDoubleu) [reply sdk/python/lib/test/automation/test_events.py]: Added it back, sorry about that

**Comment by JayDoubleu** [comment]
Hi @tgummerer i'm just following up on this. Wondering if you're happy with it. 
PS. I'm not an expert in go so if you prefer to address https://github.com/pulumi/pulumi/issues/21713 differently then I'm also happy with it and we could drop this PR. I'm just having to monkeypatch this in python on my side atm and I would prefer not to. Thanks
